builtin/providers/oneandone/resource_oneandone_firewall_policy.go

```diff
@@ -0,0 +1,299 @@
+package oneandone
+
+import (
+	"fmt"
+	"github.com/1and1/oneandone-cloudserver-sdk-go"
+	"github.com/hashicorp/terraform/helper/schema"
+	"strings"
+)
+
+func resourceOneandOneFirewallPolicy() *schema.Resource {
+	return &schema.Resource{
+
+		Create: resourceOneandOneFirewallCreate,
+		Read:   resourceOneandOneFirewallRead,
+		Update: resourceOneandOneFirewallUpdate,
+		Delete: resourceOneandOneFirewallDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"rules": {
+				Type: schema.TypeList,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"protocol": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"port_from": {
+							Type:     schema.TypeInt,
+							Optional: true,
+							ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+								if v.(int) < 1 && v.(int) > 65535 {
+									errors = append(errors, fmt.Errorf("Port end range must be between 1 and %s", "65535"))
+								}
+								return
+							},
+						},
+						"port_to": {
+							Type:     schema.TypeInt,
+							Optional: true,
+							ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+								if v.(int) < 1 && v.(int) > 65535 {
+									errors = append(errors, fmt.Errorf("Port end range must be between 1 and %s", "65535"))
+								}
+								return
+							},
+						},
+						"source_ip": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+				Required: true,
+			},
+		},
+	}
+}
+
+func resourceOneandOneFirewallCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	req := oneandone.FirewallPolicyRequest{
+		Name: d.Get("name").(string),
+	}
+
+	if desc, ok := d.GetOk("description"); ok {
+		req.Description = desc.(string)
+	}
+
+	req.Rules = getRules(d)
+
+	fw_id, fw, err := config.API.CreateFirewallPolicy(&req)
+	if err != nil {
+		return err
+	}
+
+	err = config.API.WaitForState(fw, "ACTIVE", 10, config.Retries)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(fw_id)
+
+	if err != nil {
+		return err
+	}
+
+	config.API.WaitForState(fw, "ACTIVE", 10, config.Retries)
+
+	return resourceOneandOneFirewallRead(d, meta)
+}
+
+func resourceOneandOneFirewallUpdate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	_, err := config.API.UpdateFirewallPolicy(d.Id(), d.Get("name").(string), d.Get("description").(string))
+	if err != nil {
+		return err
+	}
+
+	if d.HasChange("rules") {
+		oldR, newR := d.GetChange("rules")
+		oldValues := oldR.([]interface{})
+		newValues := newR.([]interface{})
+		if len(oldValues) > len(newValues) {
```

Comment:
Is this logic correct?

If we have oldValues, shouldn't we just range those and remove them rather than comparing the number against the newValues?